### PR TITLE
Add possibility to use TypeScript [MAILPOET-2658]

### PR DESCRIPTION
**Pull request by amine-mp**
[MAILPOET-2658]

Please check [this post on Slack](https://mailpoet.slack.com/files/T02AL75TW/FUC1MFL2F?origin_team=T02AL75TW) for details about this PR.

[MAILPOET-2658]: https://mailpoet.atlassian.net/browse/MAILPOET-2658

### CODE REVIEW

File: assets/js/src/notices/notice.tsx [hunk 1]
```diff
@@ -1,9 +1,19 @@
-import React from 'react';
+import React, { FC, ReactNode } from 'react';
 import ReactDOM from 'react-dom';
-import PropTypes from 'prop-types';
 import MailPoet from 'mailpoet';
 
-const Notice = ({
+type Props = {
+  type: 'success' | 'info' | 'warning' | 'error';
+  children: ReactNode;
```
Comment (JanJakes) [inline assets/js/src/notices/notice.tsx]: This was string/element/array-of-elements originally. Is ReactNode sufficient? Thinking about it, maybe it is - ReactNode can contain any children, right? Just double checking.

File: assets/js/src/notices/api_errors_notice.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import React, { FC } from 'react';
+import Notice from 'notices/notice';
+
+type Props = {
+  errors: Array<{ message: string }>
+}
+
+const APIErrorsNotice: FC<Props> = ({ errors }) => {
```
Comment (JanJakes) [inline assets/js/src/notices/api_errors_notice.tsx]: Is it necessary to import and use `FC` in all functional components now? Or can we do something like this (https://github.com/milesj/babel-plugin-typescript-to-proptypes):
```
const APIErrorsNotice: (props: Props) => {
  const errors = props.errors;
```

Or is there a difference? Just trying to figure out what will be the best practices now.

File: .eslintrc.ts.json [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  },
+  "plugins": [
+    "react-hooks",
+    "no-only-tests",
+    "@typescript-eslint"
+  ],
+  "settings": {
+    "import/resolver": "webpack"
+  },
+  "rules": {
```
Comment (JanJakes) [inline .eslintrc.ts.json]: These rules are something we want to turn on, eventually?

File: assets/js/src/notices/notice.tsx [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
+  scroll?: boolean;
+  closable?: boolean;
+  renderInPlace?: boolean;
+  onDisplay?: () => void;
+  onClose?: () => void;
+  timeout?: number | false;
+};
+
+const Notice: FC<Props> = ({
```
Comment (JanJakes) [inline assets/js/src/notices/notice.tsx]: Same question if we need to use `FC` here or if there's other syntax. Also... is it really nicer to unwrap the properties to such a large number of variables? Kind of thinking if `props.onClose`... isn't more convenient.

File: assets/js/src/notices/notice.tsx (hunk 1 above)
Comment (amine-mp) [reply assets/js/src/notices/notice.tsx]: Yes!

File: assets/js/src/notices/api_errors_notice.tsx (hunk 2 above)
Comment (amine-mp) [reply assets/js/src/notices/api_errors_notice.tsx]: The two ways are equivalent, the only difference is that using `FC<Props>` will implicitly add an optional `children` property of type `ReactNode`.

File: .eslintrc.ts.json (hunk 3 above)
Comment (amine-mp) [reply .eslintrc.ts.json]: Yes, we want to decide which ones to turn on and do it later.

File: assets/js/src/notices/notice.tsx (hunk 4 above)
Comment (amine-mp) [reply assets/js/src/notices/notice.tsx]: The props were spread like that before this change. But yeah I would prefer to use `props.*` in this case. Do you think I should make the change in this PR?

**Pull request merged**